### PR TITLE
DRAFT: Bounds checks

### PR DESCRIPTION
**Pull request by hanno-becker**
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

### CODE REVIEW

File: mlkem/debug.c
Comment (mkannwischer) [inline mlkem/debug.c:1]: Can we move this to a separate directory? 
I'd prefer not mix this in with the mlkem implementation.

Comment (hanno-becker) [reply mlkem/debug.c:1]: Good idea!

Comment (rod-chapman) [reply mlkem/debug.c:1]: Why are we doing this? Isn't all this runtime bounds checking rendered redundant by proof of type safety? I am all for a "belt and braces" approach but these feel like repeating ourselves - we're going to end up with preconditions that sat
 "x must be in a .. b"
immediately followed by a runtime assertion that says 
 "x must be in a .. b"

What is the proposed discipline here?

Comment (hanno-becker) [reply mlkem/debug.c:1]: > I am all for a "belt and braces" approach but these feel like repeating ourselves 

Isn't this what "belt and braces" is about? Yes, _once_ we prove the respective conditions in CBMC and/or on the ASM, they are redundant. But we are expecting to work with a variety of (ASM) implementations in the future, and debug-assertions help us catch bounds violations early in the process.

**Comment by hanno-becker** [comment]
Has been integrated into another PR.

**Pull request closed without merging**